Make hybrid annotations the default

philippotto opened the pull request:
- by default, the "New Hybrid Tracing" will be shown in the dataset list (will create a hybrid with segmentation fallback)
- a dropdown menu offers more options (e.g., skeleton, volume etc.)
- if the current dataset, does not have a segmentation layer, the fallback options are not shown
- if a segmentation layer exists, but is uint64, the default action will create an annotation without fallback data and in the dropdown the fallback-option is disabled (see #4852)
- the "create annotation" button when viewing a dataset simply defaults to creating a hybrid (with fallback if possible)
- [x] I also refactored the flow types around the dataset so that `isActive === true` always means that `dataLayers` is an array. however, the back-end also needs to be adapted so that this is really true.
- [x] remove the workaround in the rest api I added to avoid errors with the previous todo

![image](https://user-images.githubusercontent.com/2486553/98949211-b1984300-24f7-11eb-838e-f5c4113cd899.png)

I think, we can change the wording later to "New Annotation" (without "Hybrid"), but let's get the users accustomed to the new UI first.

### URL of deployed dev instance (used for testing):
- https://refactornewannotationlinks.webknossos.xyz

### Steps to test:
- use the create annotation UI in dashboard and when viewing a dataset

### Issues:
- fixes #3433

------
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Needs datastore update after deployment
- [X] Ready for review

daniel-wer reviewed (commented):
Very cool, also the flow type refactoring :+1: I only left some very small suggestions regarding the code.

Regarding the UI, I agree with some of the points @fm3 already mentioned elsewhere and think there is some room for improvement. These are a couple of notes and ideas (feel free to ignore!):
- [ ] The dropdown is misleading as it's not recognizable that only the arrow triggers the dropdown. Initially, I thought the button needed to be clicked to trigger the dropdown, however, that started a hybrid annotation. Usually one would solve this by separating the arrow:
  ![dropdown](https://user-images.githubusercontent.com/1702075/98942094-cbcd2380-24ed-11eb-906c-5adfd70ac949.png)
However, as we don't have the borders this doesn't really work :thinking: Sorry for not having a solution here, but I think we should try to find one.
- [X] A simple plus could work as an icon for the hybrid annotation
- [X] While we're at it, a cog wheel icon would work better for the "Edit" option in my opinion and one could also name it "Settings" instead of "Edit"
- [x] I think it would be fine to go with New Annotation directly instead of changing it again after a while, but no strong opinion

fm3 commented:
> - The dropdown is misleading as it's not recognizable that only the arrow triggers the dropdown. Initially, I thought the button needed to be clicked to trigger the dropdown, however, that started a hybrid annotation. Usually one would solve this by separating the arrow. However, as we don't have the borders this doesn't really work thinking Sorry for not having a solution here, but I think we should try to find one.

How about something like this, with a vertical bar and a bit more space? 
![image](https://user-images.githubusercontent.com/1390035/98945628-0e452f00-24f3-11eb-8aa1-2725ce9ac33a.png)


> - A simple plus could work as an icon for the hybrid annotation
> - While we're at it, a cog wheel icon would work better for the "Edit" option in my opinion and one could also name it "Settings" instead of "Edit"

I like that!

philippotto commented on frontend/javascripts/oxalis/model/accessors/dataset_accessor.js:
```diff
@@ -412,7 +412,7 @@ export function getDatasetExtentAsString(
+  if (!dataset.isActive) {
     return "";
   }
   const importedDataset = ((dataset: any): APIDataset);
```
Yes, that works :tada: Good catch!

philippotto commented on frontend/javascripts/oxalis/model/accessors/dataset_accessor.js:
```diff
@@ -507,7 +507,11 @@ export function isColorLayer(dataset: APIDataset, layerName: string): boolean {
+    return null;
+  }
+
   // $FlowIssue[incompatible-type]
   // $FlowIssue[prop-missing]
```
That was my initial thought, too, but the actual errors are still caused by casting the filter output to `Array<APISegmentationLayer>`. I guess, that the access to `dataset.dataSource.dataLayers` was also suppressed (and now that's not necessary, anymore), but the actual comments are still needed :/

daniel-wer replied:
Ah, I see, thanks for following up :+1:

philippotto commented:
Thank you both for your feedback and ideas! I incorporated all the suggestions and regarding the dropdown/click target issue, I decided to make the entire "New Annotation" link a dropdown menu. One can directly click on it to create a new hybrid, but already hovering over it will show the dropdown with the other options. I think, this is the most idiomatic way and also antd's default menus support this. The fact that I had to manually tinker with the CSS before should have been a sign for me, that I did something weird :see_no_evil: 

I updated the screenshot in the PR description (or check it out live).

fm3 commented:
Ah yes that’s also a fair idea :D 
I’m a bit worried, though, that this may make the original “main” button less attractive, if the user’s eye is immediately drawn to the much bigger drop down menu. Now I can imagine the user thinking the main button to be only a target to pop out the menu 🙈

philippotto commented:
> I’m a bit worried, though, that this may make the original “main” button less attractive, if the user’s eye is immediately drawn to the much bigger drop down menu. Now I can imagine the user thinking the main button to be only a target to pop out the menu

Argh, of course you are right, point taken. My first thought was to add redundancy (so the first menu item would be equal to the link), but now I'm considering the following:

![image](https://user-images.githubusercontent.com/2486553/98951839-05585b80-24fb-11eb-988b-71ebb843c343.png)

What do you think?

fm3 commented:
Yeah definitely better 🤔 What do you think @daniel-wer?

daniel-wer commented:
> I decided to make the entire "New Annotation" link a dropdown menu. One can directly click on it to create a new hybrid, but already hovering over it will show the dropdown with the other options. I think, this is the most idiomatic way and also antd's default menus support this.

I also thought about this but refuted the idea in my head with the argument @fm3 gave. However, I think you solved this nicely with the "Other options" divider. It's not perfect, but I think it's the best idea we have for now :)

philippotto commented:
Hmm, now, that I slept over it, I got more hesitant. I see two downsides:

- as @fm3 pointed out, the dropdown is quite prominent even though we actually want to encourage the other options
- the dropdown won't work on touch systems as there is no hover.

the second point is not super dramatic in my opinion, but in combination with the first, it makes me really doubt the approach.


> How about something like this, with a vertical bar and a bit more space?
> ![image](https://user-images.githubusercontent.com/1390035/98945628-0e452f00-24f3-11eb-8aa1-2725ce9ac33a.png)

I'll play around with that to see how this feels :)

philippotto commented:
Ok, I eventually went with your suggestion, @fm3. It currently looks like this:

![image](https://user-images.githubusercontent.com/2486553/100728919-7091a100-33c8-11eb-856a-1d9b131af920.png)

This solution doesn't put too much emphasis on the dropdown, the UI should be easy to understand and it also works on mobile. Seems like the right way to go :)

fm3 commented:
Great :) Thanks for looking into that! 

I noticed that on some datasets I get the skeleton-only option twice? :thinking: 
![image](https://user-images.githubusercontent.com/1390035/100731202-5c9b6e80-33cb-11eb-8e67-588738c618a3.png)

Also, do you think it might make sense to offer the same options for the create-annotation button in the menu bar in view mode?

philippotto commented:
> I noticed that on some datasets I get the skeleton-only option twice? 
> ![image](https://user-images.githubusercontent.com/1390035/100731202-5c9b6e80-33cb-11eb-8e67-588738c618a3.png)

Good catch! Should be fixed now.

> Also, do you think it might make sense to offer the same options for the create-annotation button in the menu bar in view mode?

Hm, strictly speaking, yes. I just tried adding that, but various problems arose which is why I wouldn't want to continue this, as the impact is rather low in my opinion :see_no_evil:

daniel-wer reviewed (approved):
Great work @philippotto and thanks @fm3 for the testing and ideas!

To me this looks good to be merged :tada: 

There is one minor issue but feel free to ignore - For some reason the vertical alignment of the 4 elements (icon + text + divider + arrow) irritates my brain, but I can't quite make out what's wrong:

![vertical_alignment](https://user-images.githubusercontent.com/1702075/100771311-95087000-33fe-11eb-95f1-800359be0498.png)

Maybe the arrow is too low, or the divider too high? As I said, feel free to ignore :see_no_evil:
